File: engine/src/main/java/com/example/mypixel/model/node/Vector2DNode.java
package com.example.mypixel.model.node;

import com.example.mypixel.model.*;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import org.springframework.lang.NonNull;
import java.util.Map;

@MyPixelNode("Vector2D")
public class Vector2DNode extends Node {

    @JsonCreator
    public Vector2DNode(
            @JsonProperty("id") @NonNull Long id,
            @JsonProperty("type") @NonNull String type,
            @JsonProperty("inputs") Map<String, Object> inputs) {
        super(id, type, inputs);
    }

    @Override
    public Map<String, Parameter> getInputTypes() {
        return ParamsMap.of(
                "x", Parameter.builder()
                        .type(ParameterType.DOUBLE)
                        .required(true)
                        .widget(Widget.INPUT)
                        .build(),
                "y", Parameter.builder()
                        .type(ParameterType.DOUBLE)
                        .required(true)
                        .widget(Widget.INPUT)
                        .build()
        );
    }

    @Override
    public Map<String, Object> getDefaultInputs() {
        return ParamsMap.of(
                "x", 0.0,
                "y", 0.0
        );
    }

    @Override
    public Map<String, Parameter> getOutputTypes() {
        return ParamsMap.of("vector2D", Parameter.required(ParameterType.VECTOR2D));
    }

    @Override
    public Map<String, String> getDisplayInfo() {
        return ParamsMap.of(
                "category", "Types",
                "description", "Creates a 2D vector",
                "color", "#FF8A65",
                "icon", "Vector2DIcon"
        );
    }

    @Override
    public Map<String, Object> exec() {
        Number x = (Number) inputs.get("x");
        Number y = (Number) inputs.get("y");

        Vector2D<Number> vector2D = new Vector2D<>(x, y);

        return Map.of("vector2D", vector2D);
    }

    @Override
    public void validate() {
    }
}

File: node/nodes/vector2d_node.py
from typing import Dict, Any
from node import Node
from vector2d import Vector2D


class Vector2DNode(Node):

    node_type = "Vector2D"

    def get_input_types(self) -> Dict[str, Dict[str, Any]]:
        return {
            "x": {
                "type": "DOUBLE",
                "required": True,
                "widget": "INPUT",
                "default": 0.0
            },
            "y": {
                "type": "DOUBLE",
                "required": True,
                "widget": "INPUT",
                "default": 0.0
            }
        }

    def get_output_types(self) -> Dict[str, Dict[str, Any]]:
        return {
            "vector2D": {
                "type": "VECTOR2D",
                "required": True
            }
        }

    def get_display_info(self) -> Dict[str, str]:
        return {
            "category": "Types",
            "description": "Creates a 2D vector",
            "color": "#FF8A65",
            "icon": "Vector2DIcon"
        }

    def exec(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
        x = inputs.get("x", 0.0)
        y = inputs.get("y", 0.0)

        if isinstance(x, str):
            try:
                x = float(x)
            except ValueError:
                pass

        if isinstance(y, str):
            try:
                y = float(y)
            except ValueError:
                pass

        vector2d = Vector2D(x, y)

        return {"vector2D": vector2d.to_dict()}

    def validate(self, inputs: Dict[str, Any]) -> None:
        pass
